### DISCUSSION TITLE
Need some help in accessing a command in firedrake

### DISCUSSION THREAD

**Discussion by A123-bansal123** [firedrakeproject/firedrake#2934] [Firedrake support]
1)Shall we use Gmsh file in .xml format in firedrake ?

2)M = Mesh('cylinder2.msh')
   facet_function = MeshFunction("size_t", M, 1)

error:
```
iredrake:WARNING OMP_NUM_THREADS is not set or is set to a value greater than 1, we suggest setting OMP_NUM_THREADS=1 to improve performance
Traceback (most recent call last):
  File "circle1.py", line 48, in <module>
    facet_function = MeshFunction("size_t", M, 1)
NameError: name 'MeshFunction' is not defined

```

**Comment by connorjward** [comment]
1. Using Gmsh is documented in the [manual](https://www.firedrakeproject.org/variational-problems.html#constructing-meshes). `.msh` file extensions are expected.
2. Firedrake does not have `MeshFunction`. Apart from defining forms with UFL, the FEniCs and Firedrake APIs are quite different so one shouldn't expect scripts to work unmodified with both. What are you trying to accomplish here?

**Comment by A123-bansal123** [comment]
1) when I try to convert the 3d mesh file  into a .msh file. It takes a lot of time at a refinement of 0.5?
3) How do we collect all facet functions of a square mesh with a hole in firedrake ? when i use it directly with a tag then it works but error doesn't decreases. I made same in fenics with a command attached below and it works correctly. How I can do similar thing there ?
as in FEnics i did like :
```
  mesh = Mesh(path + '.' +  'xml')
  facetfunction = MeshFunction("size_t", mesh, path + "_facet_region.xml")
   meshes.append( mesh )
   facet_functions.append(facetfunction)
and then facet_functions = boundary markers and use these boundary markers as 
ds = Measure('ds', domain=mesh, subdomain_data=boundary_markers)
bc = DirichletBC(W.sub(0), u_bc, boundary_markers, 3)
as on different refinements no of facets increase.
```
3) Below I attached a smaller version of my code. can you tell me a simpler way for the refinement of meshes(uniform or local) both. as i made change in the fourth coordinate .geo file and then generate different .msh file. In the code 3 represent the square boundary or 1 represent the circle boundary. Is there any need to use facet_function? which h is used while computing convergence rate? 
```
### code using potential flow 
from firedrake import *

#M = Mesh('cylinder2.msh')
#M = Mesh('cylinder4.msh')
M = Mesh('cylinder8.msh')
#M = Mesh('cylinder16.msh')
#M = Mesh('cylinder32.msh')

# Define Function Spaces
V = VectorFunctionSpace(M, "Lagrange", 2)
W = FunctionSpace(M, "Lagrange", 1)
Z = V * W 
# Define Test and trial function
up = Function(Z)
u, p  = split(up)
v, q = TestFunctions(Z)

nu=1
beta=-2
gamma=10
# Define normal and tangent vector
#n= FacetNormal(M)
n=  as_vector((-x,-y))
tau = as_vector([n[1], -n[0]])
h = project(CellDiameter(M), FunctionSpace(M, "DG", 0)).dat.data_ro.min()
def D(u):
    return grad(u) + grad(u).T

F = ( nu*0.5 * inner(D(u), D(v)) * dx  - Constant(nu)*dot(D(u)*dot(v,n)*n, n)*ds(1) - Constant(nu)*dot(D(v)*dot(u,n)*n, n)*ds(1) + inner(grad(u)*u, v) * dx  - p * div(v) * dx - div(u) * q * dx + p*dot(v,n)*ds(1) +q*dot(u,n)*ds(1)  - dot(f, v) * dx)
bcs = DirichletBC(Z.sub(0), u_exact, 3)

nullspace = MixedVectorSpaceBasis(Z, [Z.sub(0), VectorSpaceBasis(constant=True)])


# solve the linear system with direct solver 

solve(F == 0, up, bcs=bcs, nullspace=nullspace)
u, p = up.subfunctions

```
Hope so these details are enough and now you will be able to help me. Thankyou in advance.

**Reply by connorjward** [reply]
The manual has a section detailing how to [specify Dirichlet boundary conditions](https://firedrakeproject.org/variational-problems.html#incorporating-boundary-conditions) to portions of the mesh. I think that should answer your question.

To produce a hierarchy of refined meshes for this sort of thing I think the standard approach is to create a [`MeshHierarchy`](https://firedrakeproject.org/firedrake.mg.html#firedrake.mg.mesh.MeshHierarchy) and then access the intermediate meshes:

```py
hierarchy = MeshHierarchy(coarse_mesh, nrefinements)
coarse_mesh = hierarchy[0]
next_coarsest_mesh = hierarchy[1]
# etc
finest_mesh = hierarchy[-1]
```
